cmd/disk-cache.go

```diff
@@ -0,0 +1,674 @@
+	var format diskCacheFormat
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// Create new format.json
+		format.Version = diskCacheBackendVersion
+		format.Format = diskCacheFormatType
+		format.Time = time.Now().UTC()
+		if formatBytes, err = json.Marshal(format); err != nil {
+			return nil, err
+		}
+		if err = ioutil.WriteFile(formatPath, formatBytes, 0644); err != nil {
+			return nil, err
+		}
+	} else {
+		if err = json.Unmarshal(formatBytes, &format); err != nil {
+			return nil, err
+		}
+		if format.Version != diskCacheBackendVersion {
+			return nil, errors.New("format not supported")
+		}
+		if format.Format != diskCacheFormatType {
+			return nil, errors.New("format not supported : " + format.Format)
+		}
+	}
+
+	tmpDir := path.Join(dir, diskCacheTmpDir)
+	dataDir := path.Join(dir, diskCacheDataDir)
+	if err = os.MkdirAll(tmpDir, 0766); err != nil {
+		return nil, err
+	}
+	if err = os.MkdirAll(dataDir, 0766); err != nil {
+		return nil, err
+	}
+	boltdbPath := pathJoin(dir, diskCacheBoltDB)
+	db, err := bolt.Open(boltdbPath, 0600, nil)
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err = tx.CreateBucket([]byte(diskCacheBoltdbBucket))
+		if err != bolt.ErrBucketExists {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	cache := &diskCache{dir, maxUsage, expiry, tmpDir, dataDir, format.Time, db, make(chan struct{})}
+
+	// Start the purging go-routine
+	go cache.purge()
+	return cache, nil
+}
+
+// Abstracts disk caching - used by the S3 layer
+type cacheObjects struct {
+	dcache *diskCache
+
+	// Object functions pointing to the corresponding functions of backend implementation.
+	GetObjectFn         func(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error)
+	GetObjectInfoFn     func(bucket, object string) (objInfo ObjectInfo, err error)
+	PutObjectFn         func(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error)
+	AnonGetObjectFn     func(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error)
+	AnonGetObjectInfoFn func(bucket, object string) (objInfo ObjectInfo, err error)
+	AnonPutObjectFn     func(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error)
+}
+
+// Common function for GetObject and AnonGetObject
+func (c cacheObjects) getObject(bucket, object string, startOffset int64, length int64, writer io.Writer, anonReq bool) (err error) {
+	GetObjectFn := c.GetObjectFn
+	GetObjectInfoFn := c.GetObjectInfoFn
+	GetObjectInfo := c.GetObjectInfo
+	if anonReq {
+		GetObjectFn = c.AnonGetObjectFn
+		GetObjectInfoFn = c.AnonGetObjectInfoFn
+		GetObjectInfo = c.AnonGetObjectInfo
+	}
+	objInfo, err := GetObjectInfo(bucket, object)
+	_, backendDown := errorCause(err).(BackendDown)
```

Comment:
create this like a function `isErrBackendDown`

Reply:
using isNetworkOrHostDown()